v2/api_error.go

```diff
@@ -0,0 +1,185 @@
+package gax
+
+// code snippet credit to @ndietz
+import (
+	"fmt"
+	"strconv"
+	"strings"
+
+	"google.golang.org/genproto/googleapis/rpc/errdetails"
+	"google.golang.org/grpc/status"
+)
+
+//ErrDetails holds the google/rpc/error_details.proto messages.
+type ErrDetails struct {
+	ErrorInfo           *errdetails.ErrorInfo
+	BadRequest          *errdetails.BadRequest
+	PreconditionFailure *errdetails.PreconditionFailure
+	QuotaFailure        *errdetails.QuotaFailure
+	RetryInfo           *errdetails.RetryInfo
+	ResourceInfo        *errdetails.ResourceInfo
+	RequestInfo         *errdetails.RequestInfo
+	DebugInfo           *errdetails.DebugInfo
+	Help                *errdetails.Help
+	LocalizedMessage    *errdetails.LocalizedMessage
+
+	//Unknown stores unidentifiable error details
+	Unknown []interface{}
+}
+
+type APIError struct {
+	err     error
+	status  *status.Status
+	details ErrDetails
+}
+
+//Details presents the error details in an APIError
+func (a *APIError) Details() ErrDetails {
+	return a.details
+}
+
+//Unwrap extracts original error
+func (a *APIError) Unwrap() error {
+	return a.err
+}
+
+//Error creates a readable representation of the APIError
+func (a *APIError) Error() string {
+	var d strings.Builder
+	d.WriteString("Error Details: ")
+
+	if a.details.ErrorInfo != nil {
+		d.WriteString("api error: name = ErrorInfo" + " reason = " + a.details.ErrorInfo.Reason +
+			" domain = " + a.details.ErrorInfo.Domain + "\n")
```

Comment:
I'm not sure we should put each thing on a new line... @codyoss should an error message be limited to one continuous line? WDYT?